hitokage-core/src/widgets/workspace.rs

```diff
@@ -187,30 +296,123 @@ fn update_workspaces(root: &gtk4::FlowBox, workspaces: &Vec<WorkspaceState>) ->
                 .unwrap()
                 .set_label(&name.clone().unwrap_or((i + 1).to_string()));
               child.set_visible(*is_visible);
-              root.unselect_child(&child);
+              flowbox.unselect_child(&child);
             }
           },
-          None => root.remove(&child),
+          None => flowbox.remove(&child),
         };
       }
       None => match workspaces.get(i) {
-        Some((name, is_focused, is_visible)) => {
+        Some((name, is_focused, _)) => {
           let label = gtk::Label::new(Some(&name.clone().unwrap_or((i + 1).to_string())));
-          let child = gtk::FlowBoxChild::new();
-          // let box_ = gtk::Box::new(gtk4::Orientation::Horizontal, 1);
-          // box_.append(&label);
-          root.append(&label);
-          child.set_visible(*is_visible);
+          flowbox.append(&label);
+          let child = &flowbox.child_at_index(i as i32).unwrap();
+          child.set_hexpand(false);
           if *is_focused {
-            root.select_child(&child)
+            flowbox.select_child(child);
           }
+          child.set_size_request(width, height);
         }
         None => break,
       },
     }
+
+    if let Some(child) = flowbox.child_at_index(i as i32) {
+      let mut workspaces_to_check_constraints = workspaces_to_check_constraints_guard.lock().unwrap();
+
+      if let Some(constraints) = workspaces_to_check_constraints.get_mut(&(i as i32)) {
+        for constraint in constraints.drain(..) {
+          // log::debug!("[Workspace] removing constraint {:?}", constraint);
+          constraints_layout.remove_constraint(&constraint);
+        }
+      }
+
+      let mut to_check_constraints: Vec<Constraint> = Vec::new();
+
+      let first_visible_or_focused = workspaces
+        .iter()
+        .position(|(_, is_focused, is_visible)| *is_visible || (!*is_visible && *is_focused));
+
+      if let Some(first) = first_visible_or_focused {
+        if i == first {
+          let constraint = Constraint::new(
+            Some(&child),
+            gtk4::ConstraintAttribute::Start,
+            gtk4::ConstraintRelation::Eq,
+            Some(flowbox),
+            gtk4::ConstraintAttribute::Start,
+            1.0,
+            0.0,
+            1000000000,
+          );
+
+          to_check_constraints.push(constraint);
+        }
+      }
+
+      let prev_visible = find_first_visible_previous_sibling(&gtk4::Widget::from(child.clone()));
+
+      if i != 0 {
+        if let Some(prev) = prev_visible {
+          let constraint = Constraint::new(
+            Some(&child),
+            gtk4::ConstraintAttribute::Start,
+            gtk4::ConstraintRelation::Eq,
+            Some(&prev),
+            gtk4::ConstraintAttribute::End,
+            1.0,
+            0.0,
+            1001001000,
+          );
+
+          to_check_constraints.push(constraint);
+        }
+      }
```

Comment:
We need to change this constraint if we are at our `halign` was set to `End`, since then we expect to go backwards.